src/operator/batch_norm-inl.h

```diff
@@ -50,20 +52,294 @@ struct BatchNormParam : public dmlc::Parameter<BatchNormParam> {
   }
 };
 
-template<typename xpu>
-class BatchNormOp : public Operator {
+/*! \brief Batch normalization operator */
+template<typename xpu, typename Dtype, typename AccType>
+class BatchNormOp : public Operator
+                  , public Callbacker<Operator> {
+  typedef ::nnvm::TShape TShape;
+  typedef ::mxnet::TBlob TBlob;
+
+  /*! \brief offset, given indices such as bn, channel, depth, row, column */
+  static inline index_t offset(const TShape& shape,
+                               const size_t *indices,
+                               const size_t indicesSize) {
+    const size_t dim = shape.ndim();
+    size_t offset = 0;
+    for (size_t i = 0; i < dim; ++i) {
+      offset *= shape[i];
+      if (indicesSize > i) {
+        offset += indices[i];
+      }
+    }
+    return offset;
+  }
+
+  /*! \brief Fast-foreach when you don't care about the position other than channel */
+  template<typename Shape, typename OnData>
+  static inline void forEachFast(const Dtype *in_data, const Shape& shape, OnData onData) {
+    const size_t dim = shape.ndim();
+    const size_t num = shape[0];
+    const size_t channels = dim > 1 ? shape[1] : 1;
+    const size_t matrixSize = shape.Size() / (channels * num);
+
+    for (size_t batchItem = 0; batchItem < num; ++batchItem) {
+      #pragma openmp for
+      for (size_t channel = 0; channel < channels; ++channel) {
+        for (size_t i = 0; i < matrixSize; ++i) {
+          onData(channel, in_data++);
+        }
+      }
+    }
+  }
+
+  /*! \brief Fast-foreach when you don't care about the position other than channel */
+  template<typename Shape, typename OnData>
+  static inline void forEachFast(Dtype *in_data, const Shape& shape,
```

Comment:
Is there a need for this function?
I think we can just pass the in_data array with correct channel offset.